### PR TITLE
WFCORE-4923,read-resource operation does not resolve expressions recursively,code…

### PR DESCRIPTION
**Pull request by swetapatra**
… and test cases fix for same

 https://issues.redhat.com/browse/WFCORE-4923

### CODE REVIEW

**Comment by swetapatra** [comment]
@jmensil,@darranl can you guys take a look at this please

**Comment by bstansberry** [comment]
Hi @swetapatra,

Sorry, I see now that AbstractOperationTestCase already has 99.9% of the setup you need to test this, it's just got a small typo (incorrect "{expr}" instead of correct "${exp}") that made what's there not usable for testing this. And then it looks like that bad syntax probably was confusing when you were trying to work out how to set this up.

If you squash https://github.com/bstansberry/wildfly-core/commit/f54a5b3c20c51e22dd0f240ddd9bd1718cf83a84 into your commit that will be the needed test.

One thing to know about the tests in this maven module -- the ModelController being tested is in the same VM as the test method, which is different from most tests in the testsuite/... maven modules here or in full WildFly. So manipulating things like system properties can be done right in the test method and the operation handlers will see those values when they execute.

**Comment by swetapatra** [comment]
Hi Brian,
I tried out your updated code but I am getting :
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.277 s <<< FAILURE! - in org.jboss.as.controller.test.GlobalOperationsTestCase
[ERROR] testRecursiveReadSubModeWithResolveExpression(org.jboss.as.controller.test.GlobalOperationsTestCase)  Time elapsed: 0.019 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[test-resolve-true]> but was:<[${expr}]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.checkRecursiveSubsystem2(GlobalOperationsTestCase.java:1029)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.testRecursiveReadSubModeWithResolveExpression(GlobalOperationsTestCase.java:968)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

**Comment by swetapatra** [comment]
Hi Brian,
I tried out your updated code but I am getting :
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.277 s <<< FAILURE! - in org.jboss.as.controller.test.GlobalOperationsTestCase
[ERROR] testRecursiveReadSubModeWithResolveExpression(org.jboss.as.controller.test.GlobalOperationsTestCase)  Time elapsed: 0.019 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[test-resolve-true]> but was:<[${expr}]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.checkRecursiveSubsystem2(GlobalOperationsTestCase.java:1029)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.testRecursiveReadSubModeWithResolveExpression(GlobalOperationsTestCase.java:968)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

**Comment by swetapatra** [comment]
Hi Brian,
I tried out your updated code but I am getting :
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.277 s <<< FAILURE! - in org.jboss.as.controller.test.GlobalOperationsTestCase
[ERROR] testRecursiveReadSubModeWithResolveExpression(org.jboss.as.controller.test.GlobalOperationsTestCase)  Time elapsed: 0.019 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[test-resolve-true]> but was:<[${expr}]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.checkRecursiveSubsystem2(GlobalOperationsTestCase.java:1029)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.testRecursiveReadSubModeWithResolveExpression(GlobalOperationsTestCase.java:968)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

**Comment by swetapatra** [comment]
Hi @bstansberry,had a query
I tried out your updated code but I am getting :
[ERROR] Tests run: 22, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.277 s <<< FAILURE! - in org.jboss.as.controller.test.GlobalOperationsTestCase
[ERROR] testRecursiveReadSubModeWithResolveExpression(org.jboss.as.controller.test.GlobalOperationsTestCase)  Time elapsed: 0.019 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[test-resolve-true]> but was:<[${expr}]>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.checkRecursiveSubsystem2(GlobalOperationsTestCase.java:1029)
	at org.jboss.as.controller.test.GlobalOperationsTestCase.testRecursiveReadSubModeWithResolveExpression(GlobalOperationsTestCase.java:968)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:55)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:137)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:107)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:83)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)

**Comment by swetapatra** [comment]
Hi guys,@bristasberry @darranl @jmensil I have fixed the test case and class to reflect the changes,all the build tests are successful now,could one of you please check/review it.

**Comment by swetapatra** [comment]
Hi guys,@bristansberry @darranl @jmensil I have fixed the test case and class to reflect the changes,all the build tests are successful now,could one of you please check/review it.

**Review by jmesnil — CHANGES_REQUESTED** [review]

File: controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java [hunk 1]
```diff
@@ -954,6 +955,20 @@ public void testReadResourceAttributesOnly() throws Exception {
...
     }
 
+    @Test
+    public void testRecursiveReadSubModeWithResolveExpression() throws Exception {
+            try {   ModelNode operation = createOperation(READ_RESOURCE_OPERATION, "profile", "profileA");
+            System.setProperty("${expr}", "test-resolve-true");
+            operation.get(RECURSIVE).set(true);
+            operation.get(RESOLVE_EXPRESSIONS).set(true);
+            ModelNode result = executeForResult(operation);
+            result.get("subsystem", "subsystem2", "expression").set(System.getProperty("${expr}"));
```
Comment (jmesnil) [inline controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java:965]: this line must be removed: you are setting the value of the `expression` attribute that will override any value returned by WildFly. This makes the test pass without any confirmation that the expression has been resolved recursively by WildFly

**Comment by jmesnil** [comment]
@swetapatra great, the test is now passing and checking that the resolution of the expression is properly done by WildFly.

Could you please remove the 2 unnecessary lines in GlobalOperationsTestCase.java and squash all your commits?
I will then be able to merge your PR.

thanks!

**Comment by swetapatra** [comment]
@jmensil removed all extra lines I suppose I added them while testing and somehow forgot to remove them

File: controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java (hunk 1 above)
Comment (swetapatra) [reply controller/src/test/java/org/jboss/as/controller/test/GlobalOperationsTestCase.java]: Done,please check,I missed removing those lines earlier as was testing it again and again.

**Comment by jmesnil** [comment]
@swetapatra everything looks good now. Could you please squash your 5 commits into a single one hand I'll merge your PR. 

thanks!

**Comment by swetapatra** [comment]
I have squashed it into one  single commit,please review and let me know if anything else is required.

**Comment by darranl** [comment]
@swetapatra I think we are missing a push from your local topic branch to GitHub, this is still showing as 7 commits.

**Comment by swetapatra** [comment]
@darranl this 
57e9262  commit seems to have it all,I squashed all my prior commits into this.

**Comment by swetapatra** [comment]
@darranl this 
57e9262  commit seems to have it all,I squashed all my prior commits into this.

**Comment by jmesnil** [comment]
superseded by #4337

**Pull request closed without merging**